psi4/src/psi4/libscf_solver/sad.cc

```diff
@@ -612,43 +675,111 @@ void SADGuess::form_gradient(SharedMatrix grad, SharedMatrix F, SharedMatrix D,
+    auto Scratch1 = std::make_shared<Matrix>("Scratch1", nbf, nbf);
+    auto Scratch2 = std::make_shared<Matrix>("Scratch2", nbf, nbf);
 
     // Form Fp = XFX
     Scratch1->gemm(true, false, 1.0, X, F, 0.0);
     Scratch2->gemm(false, false, 1.0, Scratch1, X, 0.0);
 
-    auto eigvals = std::make_shared<Vector>("Eigenvalue scratch", norbs);
-    Scratch2->diagonalize(Scratch1, eigvals);
+    // Diagonalize
+    Scratch2->diagonalize(Scratch1, E);
 
     // Form C = XC'
     C->gemm(false, false, 1.0, X, Scratch1, 0.0);
 
     // Copy over Cocc
     double** Coccp = Cocc->pointer();
     double** Cp = C->pointer();
-    for (int i = 0; i < norbs; i++) {
+    for (int i = 0; i < nbf; i++) {
         C_DCOPY(nocc, Cp[i], 1, Coccp[i], 1);
     }
     // Scale by occ
     for (int i = 0; i < nocc; i++) {
-        C_DSCAL(norbs, occ->get(i), &Coccp[0][i], nocc);
+        C_DSCAL(nbf, occ->get(i), &Coccp[0][i], nocc);
     }
     // Form D = Cocc*Cocc'
     D->gemm(false, true, 1.0, Cocc, Cocc, 0.0);
 
     Scratch1.reset();
     Scratch2.reset();
 }
+SharedMatrix SADGuess::huckel_guess() {
+    // Build Neutral D in AO basis (block diagonal)
+    SharedMatrix DAO;
+    // Huckel matrices
+    SharedMatrix Chu;
+    SharedVector Ehu;
+    run_atomic_calculations(DAO, Chu, Ehu);
+
+    IntegralFactory integral(basis_, basis_, basis_, basis_);
+    MatrixFactory mat;
+
+    int nbf = basis_->nbf();
+    mat.init_with(1, &nbf, &nbf);
+    OneBodyAOInt* S_ints = integral.ao_overlap();
+
+    // Compute overlap S
+    SharedMatrix S(mat.create_matrix("Overlap Matrix"));
+    S_ints->compute(S);
+    delete S_ints;
+
+    // Compute Huckel basis overlap S*Chu
+    int nhu = Chu->coldim();
+
+    // Compute S*Chu
+    auto SChu = std::make_shared<Matrix>("SChu", nbf, nhu);
+    SChu->gemm(false, false, 1.0, S, Chu, 0.0);
+
+    // Compute Chu^T*S*Chu
+    auto ChuSChu = std::make_shared<Matrix>("ChuSChu", nhu, nhu);
+    ChuSChu->gemm(true, false, 1.0, Chu, SChu, 0.0);
+
+    // Huckel matrix in Huckel basis
+    auto huckelmo = std::make_shared<Matrix>("Huckel MO matrix", nhu, nhu);
+    double** huckelmop = huckelmo->pointer();
+    for (int i = 0; i < nhu; i++) {
+        huckelmo->set(i, i, Ehu->get(i));
+        for (int j = 0; j < nhu; j++) {
+            huckelmo->set(i, j, 0.875 * ChuSChu->get(i, j) * (Ehu->get(i) + Ehu->get(j)));
+        }
+    }
+
+    // Half-transform
+    auto scratch = std::make_shared<Matrix>("Scratch memory", nbf, nhu);
+    scratch->gemm(false, false, 1.0, SChu, huckelmo, 0.0);
+
+    // Full transform to AO basis
+    auto huckelao = std::make_shared<Matrix>("Huckel AO matrix", nbf, nbf);
+    huckelao->gemm(false, true, 1.0, SChu, scratch, 0.0);
+
+    // Now, transform from AO to SO basis
+    auto huckel = std::make_shared<Matrix>("Huckel SO matrix", AO2SO_->colspi(), AO2SO_->colspi());
+
+    auto* temp = new double[AO2SO_->rowspi()[0] * (size_t)AO2SO_->max_ncol()];
```

Comment:
This is a good place to use a `std::vector<double> temp` and in the loop below pass the pointer via `temp.data()`.

Reply:
Actually, it turns out that the SAD routine I copied this from was doing it too complicated; the Matrix class already implements the symmetry transform.